pkg/provider/ibm/provider.go

```diff
@@ -0,0 +1,183 @@
+	sm "github.com/IBM/secrets-manager-go-sdk/secretsmanagerv1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	esv1alpha1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1alpha1"
+	"github.com/external-secrets/external-secrets/pkg/provider"
+	"github.com/external-secrets/external-secrets/pkg/provider/schema"
+	"github.com/external-secrets/external-secrets/pkg/utils"
+)
+
+const (
+	SecretsManagerEndpointEnv = "IBM_SECRETSMANAGER_ENDPOINT"
+	STSEndpointEnv            = "IBM_STS_ENDPOINT"
+	SSMEndpointEnv            = "IBM_SSM_ENDPOINT"
+
+	errIBMClient                             = "cannot setup new ibm client: %w"
+	errIBMCredSecretName                     = "invalid IBM SecretStore resource: missing IBM APIKey"
+	errUninitalizedIBMProvider               = "provider IBM is not initialized"
+	errInvalidClusterStoreMissingSKNamespace = "invalid ClusterStore, missing namespace"
+	errFetchSAKSecret                        = "could not fetch SecretAccessKey secret: %w"
+	errMissingSAK                            = "missing SecretAccessKey"
+	errJSONSecretUnmarshal                   = "unable to unmarshal secret: %w"
+)
+
+type SecretManagerClient interface {
+	GetSecret(getSecretOptions *sm.GetSecretOptions) (result *sm.GetSecret, response *core.DetailedResponse, err error)
+}
+
+type providerIBM struct {
+	IBMClient SecretManagerClient
+}
+
+type client struct {
+	kube        kclient.Client
+	store       *esv1alpha1.IBMProvider
+	namespace   string
+	storeKind   string
+	credentials []byte
+}
+
+func (c *client) setAuth(ctx context.Context) error {
+	credentialsSecret := &corev1.Secret{}
+	credentialsSecretName := c.store.Auth.SecretRef.SecretAPIKey.Name
+	if credentialsSecretName == "" {
+		return fmt.Errorf(errIBMCredSecretName)
+	}
+	objectKey := types.NamespacedName{
+		Name:      credentialsSecretName,
+		Namespace: c.namespace,
+	}
+
+	// only ClusterStore is allowed to set namespace (and then it's required)
+	if c.storeKind == esv1alpha1.ClusterSecretStoreKind {
+		if c.store.Auth.SecretRef.SecretAPIKey.Namespace == nil {
+			return fmt.Errorf(errInvalidClusterStoreMissingSKNamespace)
+		}
+		objectKey.Namespace = *c.store.Auth.SecretRef.SecretAPIKey.Namespace
+	}
+
+	err := c.kube.Get(ctx, objectKey, credentialsSecret)
+	if err != nil {
+		return fmt.Errorf(errFetchSAKSecret, err)
+	}
+
+	c.credentials = credentialsSecret.Data[c.store.Auth.SecretRef.SecretAPIKey.Key]
+	if (c.credentials == nil) || (len(c.credentials) == 0) {
+		return fmt.Errorf(errMissingSAK)
+	}
+	return nil
+}
+
+func (ibm *providerIBM) GetSecret(ctx context.Context, ref esv1alpha1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	if utils.IsNil(ibm.IBMClient) {
+		return nil, fmt.Errorf(errUninitalizedIBMProvider)
+	}
+	response, _, err := ibm.IBMClient.GetSecret(
+		&sm.GetSecretOptions{
+			SecretType: core.StringPtr(sm.GetSecretOptionsSecretTypeArbitraryConst),
```

Comment:
must the `secret_type` match with the secret id? E.g. what happens if we request `type=arbitrary` when the secret is actually a `type=imported_cert` ?

Reply:
It has to match. If we request a `type=arbitrary` and we have a `type=imported_cert` on the SM, it will fail with `key \"x-x-x-x\" from ExternalSecret \"example\": Not Found"`. I would say we only support `type=arbitrary` for now. 

I am going to update the docs.